### ISSUE TITLE
add association between bytecode and line number

### ISSUE THREAD

**Issue by LFSaw** [supercollider/supercollider#3074]
such that

+ a backtrace can print the file and line number of each frame
+ it is possible to highlight currently active code
+ objects can be visually/programmaticality linked to their source code.

came up in a discussion on cosmetic improvements in error printing #2922

**Comment by jamshark70** [comment]
> it is possible to highlight currently active code

I'm not sure this would be meaningful until we have step-trace debugging -- which we would all love to have, but nobody has a concrete plan how to do it.

> objects can be visually/programmaticality linked to their source code.

While this would be useful, I can't imagine how this would be done. It all hinges on the definition of an object's "place" in the source code. I think, when you try to pin that down, it's going to be very slippery.

**Comment by brianlheim** [comment]
I'm inclined to close this on the basis of being too time-consuming to implement at the current stage of development. Unfortunately we just can't get to every great idea out there.

**Comment by LFSaw** [comment]
> I'm inclined to close this 

maybe we can still discuss a little if it is actually useful? It seems to be at least an issue people are considering to be... :)
Personally I am not particularly biased.

**Comment by brianlheim** [comment]
> maybe we can still discuss a little if it is actually useful?

The test for an issue isn't whether it's useful, it's whether we can reasonably fix it. And the answer as far as I'm concerned is no, because it would take a lot of effort we don't have and depends on first having a debugger, which is one of our long-term dreams but not in the current dev path.

**Comment by telephon** [comment]
Ah I didn't know this. Where do we put issues that we can't fix at the moment? Should there be a wiki page for that?

**Comment by brianlheim** [comment]
> Where do we put issues that we can't fix at the moment? 

We already have an entry for "make debugging easier" in https://github.com/supercollider/supercollider/projects/2. And that card also references other great ideas that will depend on an interactive debugger.

**Comment by telephon** [comment]
OK, makes sense. But there is not much space on these cards to formulate ideas. Those which do so link to a PR.

**Comment by brianlheim** [comment]
@telephon the alternative is to have a lot of open issues, many of which are "discussions" or idea boards for things that will never happen. That has the downside of dispersing dev motivation and energy. I think overall that will lead to a less productive workflow. We have seen work on PRs increase healthily after closing those which were stale.

The issue tracker is a tool for developer activity first and foremost. We have no other tool for this task. So moving nebulous and complex ideas onto the tracker for discussion is disruptive imo. Here, I see nothing that is actionable even with extended discussion and planning, so I do not think this is the place for it.

**Comment by brianlheim** [comment]
this is probably a procedural question for the dev list. sorry that i've pushed this OT

**Comment by telephon** [comment]
I agree that the dev list is better, perhaps in combination with the wiki.

**Comment by LFSaw** [comment]
> The issue tracker is a tool for developer activity first and foremost. We have no other tool for this task. So moving nebulous and complex ideas onto the tracker for discussion is disruptive imo. Here, I see nothing that is actionable even with extended discussion and planning, so I do not think this is the place for it.

I have the feeling that you have a very distinct and elaborate idea about what to discuss how and where, Brian. Others might not. 

The difficulty I see with discussions on the dev list is that they easily get buried over time. The problem with a wiki is that it is a completely unstructured environment which prevents one from finding information easily. IMHO information does not get lost there, it is lost by default.

I like the interface of the issue tracker, because it helps to, well, keep track of issues. I do not see "issue" as something related to development necessarily but more as something that is an issue, something to think about, for the development of (in this case) SuperCollider as a language, as a program, as a community.

The tracker here integrates to what github calls "projects" (i.e. message boards to collect stuff), issues can be labeled, there is easy referencing, adding a screenshot is a matter of dragging it here, there's markdown support, it is possible to _filter_ things programmatically, etc. 

Therefore I do not see so much a problem in having parts of the issue tracker that are not of interest to me, which I will likely not read  since I can programmatically filter it out (in difference to the mailinglist where I am confronted with a continuous stream of information, information getting lost, and a wiki where I have no idea about if and when something got proposed and, whether or not, in case I find information, it is already outdated since 10 years).

--- 

I don't really care if this issue here keeps on being open or not, what I do care about is that it is possible to find later on when the issue that previously was not pressing becomes interesting again. Then, things do not have to be gone through over and over again.

**Comment by telephon** [comment]
It is simply a lack in github, so we might have to improvise a little!

We do have a lot of closed pull requests and issues that are on the projects page for later. So currently, the practice is to keep closed items as threads with details for longer term projects.

@brianlheim would it be a reasonable practice to just make an issue and close it in such a case?

**Comment by brianlheim** [comment]
> The tracker here integrates to what github calls "projects" (i.e. message boards to collect stuff), issues can be labeled, there is easy referencing, adding a screenshot is a matter of dragging it here, there's markdown support, it is possible to filter things programmatically, etc.

You are totally right about that @LFSaw . It's not that I want to tell people what to discuss and where, it's that I want to use the tools we have in a way that keeps things organized and understandable, so that development work can happen efficiently. The issue tracker is very useful, and we should make the most we can of it.

> It is simply a lack in github, so we might have to improvise a little!

@telephon you've hit the nail on the head there :)

How about we mark non-actionable items with a `discussion` tag? Then we can use the issue tracker as both a discussion board and an actual issue tracker without intermingling those use cases. As you said, it would be easy to filter those issues out once they are marked. :) We don't have to close anything or rely on the dev list or wiki to archive discussions. I agree that those alternatives have their own problems, as well as adding the general downside of splitting up where discussions happen.

**Comment by brianlheim** [comment]
and maybe a `meta`, `meta-issue`, `meta-feature`, `large-scale`, etc., tag would also be useful for situations like #2943 and #2927 and #2692 ? i would love the ability to tag things that are not specific problems that can be solved through one or two PRs, but have the potential to spin out multiple PRs or issues. there's a demonstrated need for having such things on the issue tracker, but sometimes I am in the mood to browse only for things that i can solve in 1 or 2 PRs (but not necessarily just `low-hanging fruit`).

**Comment by jamshark70** [comment]
Back to topic:

> it would take a lot of effort we don't have and depends on first having a debugger

I'd say it the other way: a debugger depends on the association between line numbers and byte codes -- I don't see how you could implement a debugger without it.

The association may be useful for stack traces, even without a full interactive debugger.

Procedural: In my old job, several factors determined an issue's priority:

- Pain (crashes got high priority; cosmetic enhancements, low);
- User interest (more users = higher priority);
- Complexity (a medium-pain issue that is highly complex was likely to be delayed, even if a lot of users wanted it);
- And a few others not worth mentioning here.

I'd say this issue is medium pain, highly interesting to users but also highly complex. It probably won't get done until some renegade developer does it individually first.

> I have the feeling that you have a very distinct and elaborate idea about what to discuss how and where, Brian. Others might not.

Possibly because basically nobody else is trying to keep track of the large number of open issues.

**Comment by LFSaw** [comment]
So, do I understand this right and, independent of its possibly complicated implementation, this would be a valuable *step towards* a debugger?

> >  I have the feeling that you have a very distinct and elaborate idea about what to discuss how and where, Brian. Others might not.
>
> Possibly because basically nobody else is trying to keep track of the large number of open issues.

This is possibly true but IMHO not really an argument in an Open Source project.

**Comment by LFSaw** [comment]
🎉🥳
